src/daft-algebra/src/boolean.rs

```diff
@@ -22,3 +24,161 @@ pub fn split_conjunction(expr: &ExprRef) -> Vec<ExprRef> {
+                                right: right_right.clone(),
+                            }),
+                        }))
+                    } else if let Expr::BinaryOp {
+                        op: Operator::Or,
+                        left: left_left,
+                        right: left_right,
+                    } = left.as_ref()
+                    {
+                        // ((x | y) & z) -> ((x & z) | (y & z))
+                        Transformed::yes(Arc::new(Expr::BinaryOp {
+                            op: Operator::Or,
+                            left: Arc::new(Expr::BinaryOp {
+                                op: Operator::And,
+                                left: left_left.clone(),
+                                right: right.clone(),
+                            }),
+                            right: Arc::new(Expr::BinaryOp {
+                                op: Operator::And,
+                                left: left_right.clone(),
+                                right: right.clone(),
+                            }),
+                        }))
+                    } else {
+                        Transformed::no(e)
+                    }
+                }
+                _ => Transformed::no(e),
+            })
+        })
+        .unwrap()
+        .data
+}
+
+/// Transform boolean expression by applying De Morgan's law + eliminate double negations recursively
+fn apply_de_morgans(expr: ExprRef) -> Transformed<ExprRef> {
+    expr.transform_down(|e| {
+        Ok(match e.as_ref() {
+            Expr::Not(ne) => match ne.as_ref() {
+                // !x -> x
+                Expr::Not(nne) => Transformed::yes(nne.clone()),
+                // !(x & y) -> ((!x) | (!y))
+                Expr::BinaryOp {
+                    op: Operator::And,
+                    left,
+                    right,
+                } => Transformed::yes(Arc::new(Expr::BinaryOp {
+                    op: Operator::Or,
+                    left: left.clone().not(),
+                    right: right.clone().not(),
+                })),
+                // !(x | y) -> ((!x) & (!y))
+                Expr::BinaryOp {
+                    op: Operator::Or,
+                    left,
+                    right,
+                } => Transformed::yes(Arc::new(Expr::BinaryOp {
+                    op: Operator::And,
+                    left: left.clone().not(),
+                    right: right.clone().not(),
+                })),
+                _ => Transformed::no(e),
+            },
+            _ => Transformed::no(e),
+        })
+    })
+    .unwrap()
+}
+
+#[cfg(test)]
+mod tests {
+    use daft_dsl::col;
+
+    use crate::boolean::{to_cnf, to_dnf};
+
+    #[test]
+    fn dnf_simple() {
+        // a & (b | c) -> (a & b) | (a & c)
+        let expr = col("a").and(col("b").or(col("c")));
+        let expected = col("a").and(col("b")).or(col("a").and(col("c")));
+
+        assert_eq!(expected, to_dnf(expr));
+    }
+
+    #[test]
```

Comment:
in future I suppose could do some rust proptest stuff here to verify they are correct but not needed for this issue imo